File: 2021W27/fienestar/pgr_42839.cpp
/*
    소수 찾기
    https://programmers.co.kr/learn/courses/30/lessons/42839
*/
#include <bits/stdc++.h>

using namespace std;

class prime_checker
{
	vector<bool> is_prime;
public:
	prime_checker(size_t max_value)
		:is_prime(max(size_t(2), max_value+1), true)
	{
		is_prime[0] = is_prime[1] = false;

		for(size_t i=2; i!=is_prime.size(); ++i)
			if(is_prime[i])
				for(size_t j=i*i; j < is_prime.size(); j += i)
					is_prime[j] = false;
	}

	bool operator()(size_t n)
	{
		return is_prime[n];
	}
};

int solution(string numbers)
{
    constexpr size_t N = 10'000'000;
    vector<bool> visited(N, false);
    prime_checker is_prime(N);
    size_t count = 0;
    size_t value = 0;

    function<void(size_t)> visit = [&](size_t i){
        if(i!=0 and not visited[value] and is_prime(value)){
            ++count;
            visited[value] = true;
        }
        for(size_t k=i; k!=numbers.size(); ++k){
            swap(numbers[i], numbers[k]);
            value = value * 10 + (numbers[i] - '0');
            visit(i+1);
            value /= 10;
            swap(numbers[i], numbers[k]);
        }
    };
    
    visit(0);
    
    return count;
}

File: 2021W27/B0neh3ad/pgr_42839.py
'''
    소수 찾기
    https://programmers.co.kr/learn/courses/30/lessons/42839
'''
def check(num, lst):
    # 소수 여부 체크 -> +1 여부 결정
    cnt = 1 if num > 1 else 0
    for i in range(2, int(num ** .5) + 1):
        if num % i == 0:
            cnt = 0
            break

    # 재귀 진행 후 소수 개수 합치기
    for i in range(10):
        if lst[i]:
            lst[i] -= 1
            cnt += check(num * 10 + i, lst)
            lst[i] += 1

    return cnt


def solution(numbers):
    digits = []
    for i in range(10):
        digits.append(numbers.count(str(i)))

    answer = 0
    for i in range(1, 10):
        if digits[i]:
            digits[i] -= 1
            answer += check(i, digits)
            digits[i] += 1

    return answer
